keras/src/quantizers/quantization_config.py

```diff
@@ -0,0 +1,213 @@
+                if self.weight_quantizer.value_range != (-8, 7):
+                    raise ValueError(
+                        "Int4QuantizationConfig requires a weight_quantizer "
+                        "with value_range=(-8, 7). Received: "
+                        f"value_range={self.weight_quantizer.value_range}"
+                    )
+
+    @property
+    def mode(self):
+        return "int4"
+
+
+@keras_export("keras.quantizers.Float8QuantizationConfig")
+class Float8QuantizationConfig(QuantizationConfig):
+    """FP8 quantization config.
+
+    FP8 mixed-precision training does not support user defined quantizers.
+    This config is only used to indicate that FP8 mixed-precision training
+    should be used.
+    """
+
+    def __init__(self):
+        super().__init__(None, None)
+
+    @property
+    def mode(self):
+        return "float8"
+
+
+def validate_and_resolve_config(mode, config):
+    """Validate and resolve quantization config.
+
+    This function validates the quantization config and resolves the mode.
+    If mode is not provided, it is inferred from the config.
+    If config is not provided, a default config is inferred from the mode.
+
+    Args:
+        mode: Quantization mode.
+        config: Quantization config.
+    """
+    # 1. Backwards Compatibility: Handle string shortcuts
+    if isinstance(config, str):
+        mode = config
+        config = None
+
+    # 2. Resolve "mode" into a Config object
+    if config is None:
+        if mode == "int8":
+            config = Int8QuantizationConfig()
+        elif mode == "int4":
+            config = Int4QuantizationConfig()
+        elif mode == "float8":
+            config = Float8QuantizationConfig()
+        elif mode == "gptq":
+            raise ValueError(
+                "For GPTQ, you must pass a GPTQConfig object explicitly."
+            )
+        else:
+            if mode is not None:
+                raise ValueError(
+                    f"Invalid quantization mode. Received: mode={mode}"
+                )
+            raise ValueError(
+                "You must provide either `mode` or `config` to `quantize`."
+            )
+    else:
+        if not isinstance(config, QuantizationConfig):
+            raise ValueError(
+                "Argument `config` must be an instance of "
+                "`QuantizationConfig`. "
+                f"Received: config={config} (of type {type(config)})"
+            )
+
+    # 3. Validation: Prevent contradictions
+    if mode is not None and config.mode != mode:
+        raise ValueError(
+            f"Contradictory arguments: mode='{mode}' but "
+            f"config.mode='{config.mode}'"
+        )
+
+    # Ensure mode is consistent
+    mode = config.mode
+
+    if mode not in QUANTIZATION_MODES:
+        raise ValueError(
+            "Invalid quantization mode. "
+            f"Expected one of {QUANTIZATION_MODES}. Received: mode={mode}"
+        )
```

Comment:
Move this up line 159. The reason is that if config is None, people will see the less specific error message line 175. We want this message instead as it has more details.

Reply:
Looks like we require a check at both the places. The first one at line 159 to avoid invalid modes passed as string, and one at the existing location to validate mode derived from the config (which may have been passed directly to the quantize call)